src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -252,6 +260,8 @@ void GcsActorScheduler::HandleWorkerLeasedReply(
                   .emplace(leased_worker->GetWorkerID(), leased_worker)
                   .second);
     actor->UpdateAddress(leased_worker->GetAddress());
+    // Make sure to connect to the client before calling GCS.
+    GetOrConnectCoreWorkerClient(leased_worker->GetAddress());
```

Comment:
I'm confused why we call GetOrConnectCoreWorkerClient here, because we will call  GetOrConnectCoreWorkerClient in CreateActorOnWorker function.

Reply:
This addresses the second race condition in the description. The problem here is that we call `CreateActorOnWorker ` is called as a callback of `Put`. This means that `DestroyActor` which calls `CancelOnWorker` can be called before `CreateActorOnWorker` is called (in my integration scenario). Since `CancelOnWorker` checks if the client is already connected or not (to GC), it is broken there. => https://github.com/ray-project/ray/blob/726c4cd6d44ca10850fc4490b8fd04d0e009390a/src/ray/gcs/gcs_server/gcs_actor_scheduler.cc#L137

There is a couple solution. 

1. We make the connection before calling `Put` => this solution.
2. Just relieve the invariant check within `OnCancelNode`, meaning instead of checking if connection is already made, we just add an if statement and erase.

Reply:
Added a better comment.

Reply:
> This addresses the second race condition in the description. The problem here is that `CreateActorOnWorker ` is called as a callback of `Put`. This means that `DestroyActor` which calls `CancelOnWorker` can be called before `CreateActorOnWorker` is actually called (in my integration scenario). Since `CancelOnWorker` checks if the client is already connected or not (to GC), it is broken there. =>
> 
> https://github.com/ray-project/ray/blob/726c4cd6d44ca10850fc4490b8fd04d0e009390a/src/ray/gcs/gcs_server/gcs_actor_scheduler.cc#L137
> 
> ```
> worker in creating map
> Put
> DestroyActor # will be broken here because of RAY_CHECK
> CreateActorOnWorker
> ```
> 
> There is a couple solution.
> 
> 1. We make the connection before calling `Put` => this solution.
> 2. Just relieve the invariant check within `OnCancelNode`, meaning instead of checking if connection is already made, we just add an if statement and erase.

As you said, the `CreateActorOnWorker` may be invoked after `DestroyActor` as the `CreateActorOnWorker` inside the callback of `Put`.
It means that the `CreateActorOnWorker` will be invoked anyway.
I think it still has problem.

Reply:
I think you could check the actor status inside `CreateActorOnWorker`, if it already died then just return.

Reply:
This part is added to avoid breaking `DestroyActor`, which calles `CancelOnNode` which checks if the client is already created to GC it. The scenario you mentioned is already covered inside `CreateActorOnWorker`.

Reply:
https://github.com/ray-project/ray/blob/f395e480316e31d47883c746f06a1b45e1bc7711/src/ray/gcs/gcs_server/gcs_actor_scheduler.cc#L287